src/tools/compiletest/src/load_cfg/itemlist.rs

```diff
@@ -0,0 +1,733 @@
+                    "if this visit has never happened before, there should be \
+                    no expanded nodes."
+                ),
+                _ => (),
+            }
+        }
+
+        Ok(())
+    }
+}
+
+// TODO: parse revisions into `ItemVal`s.
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum ItemVal<'src> {
+    /* Flags that can only be set to true */
+    BuildPass,
+    BuildFail,
+    CheckPass,
+    CheckFail,
+    RunPass,
+    RunFail,
+    NoPreferDynamic,
+    NoAutoCheckCfg,
+
+    /* Boolean flags that can be set with `foo-bar` or `no-foo-bar` */
+    ShouldIce(bool),
+    ShouldFail(bool),
+    BuildAuxDocs(bool),
+    ForceHost(bool),
+    CheckStdout(bool),
+    CheckRunResults(bool),
+    DontCheckCompilerStdout(bool),
+    DontCheckCompilerStderr(bool),
+    PrettyExpanded(bool),
+    PrettyCompareOnly(bool),
+    CheckTestLineNumbersMatch(bool),
+    StderrPerBitwidth(bool),
+    Incremental(bool),
+    DontCheckFailureStatus(bool),
+    RunRustfix(bool),
+    RustfixOnlyMachineApplicable(bool),
+    CompareOutputLinesBySubset(bool),
+    KnownBug(bool),
+    RemapSrcBase(bool),
+
+    /* Key-value directivesthat can be set with `foo-bar: baz qux` */
+    ErrorPattern(&'src str),
+    RegexErrorPattern(&'src str),
+    CompileFlags(&'src str),
+    Edition(&'src str),
+    RunFlags(&'src str),
+    PrettyMode(&'src str),
+    AuxBin(&'src str),
+    AuxBuild(&'src str),
+    AuxCrate(&'src str),
+    AuxCodegenBackend(&'src str),
+    ExecEnv(&'src str),
+    UnsetExecEnv(&'src str),
+    RustcEnv(&'src str),
+    UnsetRustcEnv(&'src str),
+    ForbidOutput(&'src str),
+    FailureStatus(&'src str),
+    AssemblyOutput(&'src str),
+    TestMirPass(&'src str),
+    LlvmCovFlags(&'src str),
+    FilecheckFlags(&'src str),
+    /// The revisions list
+    Revisions(&'src str),
+
+    /* Directives that are single keys but have variable suffixes */
+    /// Ignore something specific about this test.
+    ///
+    /// `//@ ignore-x86`
+    Ignore {
+        what: &'src str,
+    },
+    /// Components needed by the test
+    ///
+    /// `//@ needs-rust-lld`
+    Needs {
+        what: &'src str,
+    },
+    /// Run only if conditions are met
+    ///
+    /// `//@ only-linux`
+    Only {
+        what: &'src str,
+    },
+    /// Normalizations make transformations in output files before checking
+    ///
+    /// `//@ normalize-stdout: foo -> bar`
+    /// `//@ normalize-stderr-32bit: bar -> baz`
+    /// `//@ normalize-stderr-test: "h[[:xdigit:]]{16}" -> "h[HASH]"`
+    Normalize {
+        what: &'src str,
+    },
+
+    /* Specialized patterns */
+    /// Directives that apply to a single revision. Don't process this here; just extract the
+    /// key and value. The content will be reprocessed as a new item.
+    ///
+    /// Note that this does not yet recurse.
+    ///
+    /// `//@[revname] directive`
+    RevisionSpecificItems {
+        revs: &'src str,
```

Comment:
Discussion: we probably want revisions to be represented as `BTreeSet<&'src str>`, maybe. Or better yet, we could try to do something similar to `ui_test` and have revision-specific directive set:

```rs
enum UiRevision {
    None,
    Named(String),
}

/// Revision specific?
struct UiDirectiveSet {
    check_pass: bool,
    edition: String,
    // ...
}

type UiDirectives = BTreeMap<UiRevision, UiDirectiveSet>;
```

Reply:
The `ui_test` form is almost exactly what I was going for, just haven't gotten to yet :) my plan is to parse all config that is not specific to revisions into one `Config`, then clone it per revision, then update the config with anything that is per revision (output lives around here https://github.com/rust-lang/rust/pull/128070/files#diff-4653d6449a3cc26e6235d71fbf2919348e09fe8eaa72db6fc8e623eaa92ce1f5R38-R43, didn't handle the 0-revision case yet).